File: desktop/src/lib/local-mandates.ts
/**
 * Local-model Mandates — the desktop's on-device intelligence, resolved by the
 * PLATFORM, run on the local llama-server.
 *
 * Every intelligence point runs through a Mandate so a Mandate change is
 * guaranteed to take effect (SoR: common-docs/systems/intelligence/mandates/STATE.md
 * §9, §10 Satellites). "Offline" is a data LOCATION, never a different
 * intelligence: the Mandate decides WHICH agent — its instructions, variables,
 * sampling settings and output schema — and this device only supplies the
 * compute. There is no prompt in this repo for these jobs.
 *
 * Flow: GET {engine}/local-mandates/{key} — the engine resolves the mandate
 * against AIDream as this person in their active organization, keeps that
 * answer (and the Holder's execution definition) in SQLite, and hands back
 * both. With the network down it serves the LAST answer the platform gave,
 * flagged `stale` with the reason (app/services/ai/local_mandates.py). Never a
 * client-side fallback agent or a seed prompt.
 */

import { engine } from "@/lib/api";
import type {
  AgentExecutionDefinition,
  MandateResolution,
} from "@/lib/aidream-client";

/**
 * The local-model Mandate keys, declared in aidream
 * `aidream/services/mandates/client_mandates.py` (2026-09-25, rounds 1 and 2).
 *
 * TODO(@ai-matrx/agents > 0.13.8): round 2's keys are regenerated into
 * `apps/shared/matrx-agents/mandates/keys.generated.ts` but not published yet
 * (installed here: 0.13.7). When the package carries them, replace each literal
 * with `MANDATE_KEYS.local__<job>` from `@ai-matrx/agents/mandates` so a rename
 * or retirement fails type-check here (the way `DEFAULT_CHAT_MANDATE_KEY` in
 * `@/lib/mandates` already does).
 */
export const LOCAL_MODEL_MANDATE_KEYS = {
  polishTranscript: "local.polish_transcript",
  summarizeText: "local.summarize_text",
  improveWriting: "local.improve_writing",
  extractActionItems: "local.extract_action_items",
  explainCode: "local.explain_code",
  answerQuestion: "local.answer_question",
  confidentialChat: "local.confidential_chat",
  /** Confidential Chat's Tools mode (the person's selected local tools). */
  toolCallingChat: "local.tool_calling_chat",
  /** Confidential Chat's Raw JSON mode (the person's hand-written request body). */
  rawCompletion: "local.raw_completion",
  /** Voice page AI Polish — built-in styles and the person's own style. */
  polishStyleStandard: "local.polish_style_standard",
  polishStyleFormal: "local.polish_style_formal",
  polishStyleBullets: "local.polish_style_bullets",
  polishStyleActionItems: "local.polish_style_action_items",
  polishStyleMeetingNotes: "local.polish_style_meeting_notes",
  polishStyleLightCleanup: "local.polish_style_light_cleanup",
  polishStyleCustom: "local.polish_style_custom",
  /** Confidential Chat's built-in prompt library (the Holder's system text). */
  chatPersonaHelpful: "local.chat_persona_helpful",
  chatPersonaTranscriptPolish: "local.chat_persona_transcript_polish",
  chatPersonaSummarize: "local.chat_persona_summarize",
  chatPersonaExplainSimply: "local.chat_persona_explain_simply",
  chatPersonaCodeReview: "local.chat_persona_code_review",
  chatPersonaBrainstorm: "local.chat_persona_brainstorm",
  chatPersonaSpokenReplies: "local.chat_persona_spoken_replies",
} as const;

export type LocalModelMandateKey =
  (typeof LOCAL_MODEL_MANDATE_KEYS)[keyof typeof LOCAL_MODEL_MANDATE_KEYS];

/** One authored message of the resolved Holder, flattened to plain text. */
export interface HolderMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

/** Sampling values the Holder declares (absent = the local server's default). */
export interface HolderSettings {
  temperature?: number;
  topP?: number;
  maxTokens?: number;
}

export interface ResolvedLocalMandate {
  mandateKey: LocalModelMandateKey;
  resolution: MandateResolution;
  definition: AgentExecutionDefinition;
  messages: HolderMessage[];
  settings: HolderSettings;
  outputSchema: Record<string, unknown> | null;
  /**
   * True when the platform was unreachable and this is the LAST answer it
   * gave this person in this organization (cached on this device).
   */
  stale: boolean;
  /** Why the answer is stale; null when fresh. */
  staleReason: string | null;
}

/**
 * Window event fired when a run used a stale (offline) answer; the app's
 * notification center announces it (hooks/use-notifications.ts).
 */
export const LOCAL_MANDATE_STALE_EVENT = "matrx-local-mandate-stale";

/** What `GET {engine}/local-mandates/{key}` answers. */
interface EngineLocalMandate {
  mandate_key: string;
  resolution: MandateResolution;
  definition: AgentExecutionDefinition;
  resolved_at: string;
  stale: boolean;
  stale_reason: string | null;
}

/**
 * A short cache so a chat turn does not pay two round trips; short enough that
 * a rebind reaches the next run within a minute. The server is the authority.
 */
const CACHE_TTL_MS = 60_000;
const cache = new Map<string, { at: number; value: ResolvedLocalMandate }>();

function textOf(content: unknown): string {
  if (typeof content === "string") return content;
  if (Array.isArray(content)) {
    return content
      .map((block) =>
        block && typeof block === "object" && "text" in block
          ? String((block as { text: unknown }).text ?? "")
          : "",
      )
      .join("");
  }
  return "";
}

function holderMessages(definition: AgentExecutionDefinition): HolderMessage[] {
  const out: HolderMessage[] = [];
  for (const raw of definition.messages) {
    if (!raw || typeof raw !== "object") continue;
    const role = (raw as { role?: unknown }).role;
    if (role !== "system" && role !== "user" && role !== "assistant") continue;
    out.push({ role, content: textOf((raw as { content?: unknown }).content) });
  }
  return out;
}

function num(value: unknown): number | undefined {
  return typeof value === "number" && Number.isFinite(value) ? value : undefined;
}

function holderSettings(definition: AgentExecutionDefinition): HolderSettings {
  const s = definition.settings ?? {};
  const out: HolderSettings = {};
  const temperature = num(s["temperature"]);
  const topP = num(s["top_p"]);
  const maxTokens = num(s["max_output_tokens"]) ?? num(s["max_tokens"]);
  if (temperature !== undefined) out.temperature = temperature;
  if (topP !== undefined) out.topP = topP;
  if (maxTokens !== undefined) out.maxTokens = maxTokens;
  return out;
}

/**
 * Resolve a local-model Mandate for THIS user and organization and load its
 * Holder's definition — through the engine, so the last answer survives
 * offline. Throws with the Mandate key named when anything fails — the caller
 * refuses the run and shows the message; it never falls back.
 */
export async function resolveLocalMandate(
  mandateKey: LocalModelMandateKey,
  signal?: AbortSignal,
): Promise<ResolvedLocalMandate> {
  const hit = cache.get(mandateKey);
  if (hit && Date.now() - hit.at < CACHE_TTL_MS) return hit.value;
  try {
    const raw = (await engine.get(
      `/local-mandates/${encodeURIComponent(mandateKey)}`,
      signal ? { signal } : undefined,
    )) as EngineLocalMandate | null;
    if (!raw || typeof raw !== "object" || !raw.resolution || !raw.definition) {
      throw new Error("the engine answered without a resolution and a Holder definition");
    }
    const definition = raw.definition;
    const messages = holderMessages(definition);
    if (messages.length === 0) {
      throw new Error(`agent ${raw.resolution.agent_id} has no authored messages`);
    }
    const value: ResolvedLocalMandate = {
      mandateKey,
      resolution: raw.resolution,
      definition,
      messages,
      settings: holderSettings(definition),
      outputSchema: definition.output_schema ?? null,
      stale: raw.stale === true,
      staleReason: raw.stale === true ? (raw.stale_reason ?? "the platform was unreachable") : null,
    };
    // A stale answer is never cached here: the next run asks again, so a
    // reconnect (or a rebind) reaches it at once.
    if (!value.stale) cache.set(mandateKey, { at: Date.now(), value });
    else if (typeof window !== "undefined") {
      window.dispatchEvent(
        new CustomEvent(LOCAL_MANDATE_STALE_EVENT, {
          detail: { mandateKey, reason: value.staleReason },
        }),
      );
    }
    return value;
  } catch (err) {
    const detail = err instanceof Error ? err.message : String(err);
    throw new Error(`Could not resolve ${mandateKey}: ${detail}`);
  }
}

/**
 * The Holder's authored messages that lead a conversation — a system message
 * with no text is dropped (it carries no instructions), everything else is
 * kept in order.
 */
export function holderLeadMessages(holder: ResolvedLocalMandate): HolderMessage[] {
  return holder.messages.filter((m) => m.role !== "system" || m.content.trim() !== "");
}

/** Replace `{{name}}` with the value when one is supplied; leave others as authored. */
export function substituteVariables(
  template: string,
  vars: Record<string, string>,
): string {
  return template.replace(/\{\{(\w+)\}\}/g, (match, key: string) =>
    Object.prototype.hasOwnProperty.call(vars, key) ? (vars[key] ?? match) : match,
  );
}

/** Test seam: forget cached resolutions. */
export function clearLocalMandateCache(): void {
  cache.clear();
}

File: app/services/ai/local_mandates.py
"""Local-model Mandates — the platform's answer, kept for offline runs.

Every on-device intelligence point in the desktop (AI Polish, the text
pipeline, Confidential Chat and its prompt library, Tools and Raw JSON modes)
runs through a Mandate: the PLATFORM decides which agent holds the job — its
instructions, variables, sampling and output schema — and this device only
supplies the compute (common-docs/systems/intelligence/mandates/STATE.md §9).

"Offline" is a data LOCATION, never a different intelligence. This module is
the ONE door the desktop asks: it resolves the mandate against AIDream as the
signed-in person in their active organization, stores that answer verbatim
(SQLite ``mandate_resolutions``), and loads the Holder's execution definition
through the existing fetch-through cache (``LocalExecutionAgentSource`` over
``agent_execution_definitions``). With the network down it serves the LAST
answer the platform gave, flagged stale with the reason — never a prompt that
lives in code, and never a silent substitute.

What it refuses (the platform's word is final when it answers):
  * AIDream answers 401/403/404/409/422 for the mandate → that refusal is
    passed through; a cached answer is NOT served over it.
  * Nothing was ever resolved on this device and the platform is unreachable
    → 503 naming the mandate and what to do.
"""

from __future__ import annotations

import logging
from typing import Any
from urllib.parse import quote

from fastapi import HTTPException, status

from app.services.ai.agent_source import get_execution_agent_source
from app.services.aidream.client import (
    AIDreamError,
    AIDreamOfflineError,
    get_aidream_client,
)
from app.services.aidream.organization import (
    OrganizationNotResolvedError,
    get_device_organization,
    resolve_active_organization_id,
)
from app.services.local_db.repositories import MandateResolutionsRepo, TokenRepo

logger = logging.getLogger(__name__)

#: AIDream statuses that are the platform's own answer about this mandate for
#: this caller. Serving a cached resolution over one of these would run an
#: agent the platform just refused.
_AUTHORITATIVE_REFUSALS = frozenset({401, 403, 404, 409, 422})


def _resolution_path(mandate_key: str) -> str:
    return f"/mandates/{quote(mandate_key, safe='')}/resolution"


def _checked_resolution(mandate_key: str, payload: Any) -> dict[str, Any]:
    if not isinstance(payload, dict):
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail={
                "code": "mandate_resolution_invalid",
                "message": f"AI Matrx answered {mandate_key} with a non-object resolution.",
            },
        )
    agent_id = payload.get("agent_id")
    if not isinstance(agent_id, str) or not agent_id:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail={
                "code": "mandate_resolution_invalid",
                "message": f"AI Matrx's resolution of {mandate_key} names no agent.",
            },
        )
    if not isinstance(payload.get("is_version"), bool):
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail={
                "code": "mandate_resolution_invalid",
                "message": f"AI Matrx's resolution of {mandate_key} has no is_version flag.",
            },
        )
    return payload


class LocalMandateResolver:
    def __init__(
        self,
        repo: MandateResolutionsRepo | None = None,
        token_repo: TokenRepo | None = None,
    ) -> None:
        self._repo = repo or MandateResolutionsRepo()
        self._token_repo = token_repo or TokenRepo()

    async def _fresh(
        self, mandate_key: str, jwt: str, user_id: str
    ) -> tuple[dict[str, Any], str, str]:
        """Ask AIDream. Returns (resolution, organization_id, fetched_at)."""
        client = get_aidream_client()
        if client is None:
            raise AIDreamOfflineError("the AI Matrx server URL is not configured")
        organization_id = await resolve_active_organization_id(jwt)
        payload = await client.get(
            _resolution_path(mandate_key),
            jwt=jwt,
            headers={"X-Organization-Id": organization_id},
        )
        resolution = _checked_resolution(mandate_key, payload)
        fetched_at = await self._repo.upsert(
            mandate_key,
            user_id=user_id,
            organization_id=organization_id,
            resolution=resolution,
        )
        return resolution, organization_id, fetched_at

    async def resolve(self, mandate_key: str) -> dict[str, Any]:
        token = await self._token_repo.get()
        jwt: str | None = None
        if token and not self._token_repo.is_expired(token):
            raw = token.get("access_token")
            jwt = raw if isinstance(raw, str) and raw else None
        user_id = (token or {}).get("user_id") or await self._token_repo.get_owner_user_id()
        if not user_id:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail={
                    "code": "authentication_required",
                    "message": (
                        f"Sign in to AI Matrx — {mandate_key}'s instructions come from "
                        "your account, and this device has never been signed in."
                    ),
                },
            )

        why_stale: str | None = None
        if jwt is None:
            why_stale = "you are signed out or your session has expired"
        else:
            try:
                resolution, organization_id, fetched_at = await self._fresh(
                    mandate_key, jwt, str(user_id)
                )
            except AIDreamOfflineError as exc:
                why_stale = f"AI Matrx is unreachable ({exc})"
            except OrganizationNotResolvedError as exc:
                why_stale = f"your organization could not be confirmed ({exc})"
            except AIDreamError as exc:
                if exc.status in _AUTHORITATIVE_REFUSALS:
                    raise HTTPException(
                        status_code=exc.status,
                        detail={
                            "code": "mandate_refused",
                            "message": f"AI Matrx refused {mandate_key}: {exc}",
                        },
                    ) from exc
                why_stale = f"AI Matrx answered HTTP {exc.status}"
            else:
                return await self._with_definition(
                    mandate_key, resolution, fetched_at, stale_reason=None
                )

        organization_id = await get_device_organization(str(user_id))
        cached = (
            await self._repo.get(
                mandate_key, user_id=str(user_id), organization_id=organization_id
            )
            if organization_id
            else None
        )
        if cached is None or not isinstance(cached.get("resolution_json"), dict):
            raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
                detail={
                    "code": "mandate_unavailable_offline",
                    "message": (
                        f"{mandate_key} has never been resolved on this device, and {why_stale}. "
                        "Connect to AI Matrx once so this device learns which instructions to run; "
                        "after that it keeps working offline."
                    ),
                },
            )
        logger.warning(
            "[local_mandates] serving the cached resolution of %s from %s — %s",
            mandate_key,
            cached.get("fetched_at"),
            why_stale,
        )
        return await self._with_definition(
            mandate_key,
            cached["resolution_json"],
            str(cached.get("fetched_at") or ""),
            stale_reason=why_stale,
        )

    async def _with_definition(
        self,
        mandate_key: str,
        resolution: dict[str, Any],
        fetched_at: str,
        *,
        stale_reason: str | None,
    ) -> dict[str, Any]:
        definition = await get_execution_agent_source().load_for_execution(
            str(resolution["agent_id"]),
            is_version=bool(resolution["is_version"]),
        )
        return {
            "mandate_key": mandate_key,
            "resolution": resolution,
            "definition": definition.model_dump(mode="json"),
            "resolved_at": fetched_at,
            "stale": stale_reason is not None,
            "stale_reason": stale_reason,
        }


_instance: LocalMandateResolver | None = None


def get_local_mandate_resolver() -> LocalMandateResolver:
    global _instance
    if _instance is None:
        _instance = LocalMandateResolver()
    return _instance


__all__ = ["LocalMandateResolver", "get_local_mandate_resolver"]
